### PR TITLE
[EC-64] Sync from multiple peers

### PR DESCRIPTION
**Pull request by LukasGasior1**
Added: FastSyncController + child actors that handle all four kinds of requests: block headers, block bodies, receipts and state nodes.

Controller tries to send requests to as many peers as possible (but within the `max-concurrent-requests` limit). Each child worker sends the request and waits for the response from peer. Syncing is done when:
there are no more queued blocks/state nodes, there are no requests waiting for the response and current block number == target block number.

When peer disappoints us (for example it does not respond or the response is empty) we mark it as blacklisted (for some time) and don't send any more requests to it.

@adamsmo as it strongly bases on your code from SyncActor, please take a closer look if I haven't missed anything :)

TODO: tests, improve performance

### CODE REVIEW

File: src/main/scala/io/iohk/ethereum/blockchain/sync/FastSyncController.scala [hunk 1]
```diff
@@ -0,0 +1,294 @@
...
+
+  case class StartFastSync(targetBlockHash: ByteString)
+
+  case class EnqueueNodes(hashes: Seq[HashType])
+  case class EnqueueBlockBodies(hashes: Seq[ByteString])
+  case class EnqueueReceipts(hashes: Seq[ByteString])
+
+  case class UpdateDownloadedNodesCount(update: Int)
+
+  trait HashType {
```
Comment (ntallar) [inline src/main/scala/io/iohk/ethereum/blockchain/sync/FastSyncController.scala]: Maybe this trait should be `sealed`? After upgrading to Scala 2.12 the compiler warns me of match not exhaustive in `handleMptDownload` if this trait isn't sealed.

**Comment by LukasGasior1** [comment]
Sure it should

File: src/main/scala/io/iohk/ethereum/blockchain/sync/FastSyncReceiptsRequestHandler.scala [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+
+  override val requestMsg = GetReceipts(requestedHashes)
+  override val responseMsgCode = Receipts.code
+
+  override def handleResponseMsg(receipts: Receipts): Unit = {
+    (requestedHashes zip receipts.receiptsForBlocks).foreach { case (hash, receiptsForBlock) =>
+      receiptStorage.put(hash, receiptsForBlock)
+    }
+
+    if (receipts.receiptsForBlocks.isEmpty) {
```
Comment (adamsmo) [inline src/main/scala/io/iohk/ethereum/blockchain/sync/FastSyncReceiptsRequestHandler.scala]: where we are removing elements from `receiptsForBlocks`

File: src/test/scala/io/iohk/ethereum/network/p2p/FastSyncActorSpec.scala [hunk 3]
```diff
@@ -7,14 +7,13 @@ import io.iohk.ethereum.crypto
...
 import io.iohk.ethereum.domain.BlockHeader
-import io.iohk.ethereum.network.FastSyncActor.{FastSyncDone, SyncFailure}
 import io.iohk.ethereum.network.PeerActor.MessageReceived
 import io.iohk.ethereum.network.p2p.messages.PV62._
 import io.iohk.ethereum.network.p2p.messages.PV63._
-import io.iohk.ethereum.network.{FastSyncActor, PeerActor}
 import org.scalatest.{FlatSpec, Matchers}
 import org.spongycastle.util.encoders.Hex
 
+/*
```
Comment (adamsmo) [inline src/test/scala/io/iohk/ethereum/network/p2p/FastSyncActorSpec.scala]: should this be removed?

File: src/main/scala/io/iohk/ethereum/App.scala [hunk 4]
```diff
@@ -35,21 +35,17 @@ object App {
...
+    val fastSyncController = actorSystem.actorOf(FastSyncController.props(
+      peerManager,
+      nodeStatusHolder,
+      new MptNodeStorage(dataSource),
+      new BlockHeadersStorage(dataSource),
+      new BlockBodiesStorage(dataSource),
+      new ReceiptStorage(dataSource),
+      new EvmCodeStorage(dataSource)), "fast-sync-controller")
+
+    fastSyncController ! FastSyncController.StartFastSync(ByteString(Hex.decode("81e2dcb132c2af3cb84591466aa904bb054f0b9ba52e369c06a271f6d92190db")))
```
Comment (LukasGasior1) [inline src/main/scala/io/iohk/ethereum/App.scala]: good point

File: src/main/scala/io/iohk/ethereum/blockchain/sync/FastSyncBlockHeadersRequestHandler.scala [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+    blockHeaders.headers.lastOption foreach { lastHeader =>
+      nodeStatusHolder.send(_.copy(
+        blockchainStatus = BlockchainStatus(lastHeader.difficulty, lastHeader.hash, lastHeader.number)))
+    }
+
+    if (blockHashes.nonEmpty) {
+      fastSyncController ! FastSyncController.EnqueueBlockBodies(blockHashes)
+      fastSyncController ! FastSyncController.EnqueueReceipts(blockHashes)
+    } else {
+      fastSyncController ! BlacklistSupport.BlacklistPeer(peer)
```
Comment (LukasGasior1) [inline src/main/scala/io/iohk/ethereum/blockchain/sync/FastSyncBlockHeadersRequestHandler.scala]: we don't have a queue for headers, we download new ones basing on our current header

File: src/main/scala/io/iohk/ethereum/blockchain/sync/FastSyncController.scala [hunk 6]
```diff
@@ -0,0 +1,301 @@
...
+        assignedHandlers -= sender()
+        processSyncing()
+
+      case Terminated(ref) if assignedHandlers.contains(ref) =>
+        context unwatch ref
+        assignedHandlers -= ref
+
+      case PrintStatus =>
+        val totalNodesCount = downloadedNodesCount + nodesQueue.size
+        log.error(
```
Comment (LukasGasior1) [inline src/main/scala/io/iohk/ethereum/blockchain/sync/FastSyncController.scala]: nope, will change to info

File: src/main/scala/io/iohk/ethereum/blockchain/sync/FastSyncController.scala [hunk 7]
```diff
@@ -0,0 +1,301 @@
...
+
+    def finish(): Unit = {
+      log.info("Fast sync finished")
+      context stop self
+    }
+
+    def processQueues(): Unit = {
+      if (unassignedPeers.isEmpty) {
+        if (assignedHandlers.nonEmpty) {
+          log.error("There are no available peers, waiting for responses")
```
Comment (LukasGasior1) [inline src/main/scala/io/iohk/ethereum/blockchain/sync/FastSyncController.scala]: nope, it's info. I changed it (and few more) to error when testing. will fix

File: src/main/scala/io/iohk/ethereum/blockchain/sync/FastSyncReceiptsRequestHandler.scala (hunk 2 above)
Comment (LukasGasior1) [inline src/main/scala/io/iohk/ethereum/blockchain/sync/FastSyncReceiptsRequestHandler.scala]: we are not removing them. it's the response from peer, we save all received receipts above. here is the check if the response was empty (then we blacklist)

File: src/test/scala/io/iohk/ethereum/network/p2p/FastSyncActorSpec.scala (hunk 3 above)
Comment (LukasGasior1) [inline src/test/scala/io/iohk/ethereum/network/p2p/FastSyncActorSpec.scala]: Yes, I'll remove it

File: src/main/scala/io/iohk/ethereum/blockchain/sync/FastSyncBlockBodiesRequestHandler.scala [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+    requestedHashes: Seq[ByteString],
+    blockBodiesStorage: BlockBodiesStorage)(implicit scheduler: Scheduler)
+  extends FastSyncRequestHandler[GetBlockBodies, BlockBodies](peer) {
+
+  override val requestMsg = GetBlockBodies(requestedHashes)
+  override val responseMsgCode = BlockBodies.code
+
+  override def handleResponseMsg(blockBodies: BlockBodies): Unit = {
+    (requestedHashes zip blockBodies.bodies).foreach { case (hash, body) =>
+      blockBodiesStorage.put(hash, body)
```
Comment (AlanVerbner) [inline src/main/scala/io/iohk/ethereum/blockchain/sync/FastSyncBlockBodiesRequestHandler.scala]: we need to double check our DataSources are thread safe now ....I can recall `LevelDB` is but IODB not...will it be better to have a single _handleBodies_ queue to push those changes to storage?

Comment (LukasGasior1) [inline src/main/scala/io/iohk/ethereum/blockchain/sync/FastSyncBlockBodiesRequestHandler.scala]: in such case we could write a thin wrapper that implements BlockBodiesStorage (and others) and handles requests from multiple threads

File: src/main/scala/io/iohk/ethereum/blockchain/sync/FastSyncController.scala [hunk 9]
```diff
@@ -0,0 +1,301 @@
...
+    case PeerActor.StatusResponse(status) =>
+      if (status == PeerActor.Status.Handshaked) {
+        if (!handshakedPeers.contains(sender()) && !isBlacklisted(sender())) {
+          handshakedPeers += sender()
+          context watch sender()
+        }
+      } else removePeer(sender())
+
+    case Terminated(ref) if handshakedPeers.contains(ref) =>
+      removePeer(ref)
```
Comment (AlanVerbner) [inline src/main/scala/io/iohk/ethereum/blockchain/sync/FastSyncController.scala]: is there any chance there are still _pending assigned_ requests to the peer (ie: block bodies) that should be returned in the same way as if how it's done when a timeout occurs?

File: src/main/scala/io/iohk/ethereum/blockchain/sync/FastSyncController.scala [hunk 10]
```diff
@@ -0,0 +1,301 @@
...
+          log.info("Peer ({}) did not respond with target block header, blacklisting and scheduling retry in {}",
+            sender().path.name,
+            startRetryInterval)
+
+          blacklist(sender(), blacklistDuration)
+          scheduleStartFastSync(startRetryInterval, targetBlockHash)
+          context become idle
+      }
+
+    case TargetBlockTimeout =>
```
Comment (AlanVerbner) [inline src/main/scala/io/iohk/ethereum/blockchain/sync/FastSyncController.scala]: :heart:

File: src/main/scala/io/iohk/ethereum/blockchain/sync/FastSyncController.scala (hunk 9 above)
Comment (LukasGasior1) [inline src/main/scala/io/iohk/ethereum/blockchain/sync/FastSyncController.scala]: If a child handler is still waiting for response from that peer it will also receive Terminated msg and do `cleanupAndStop` (see FastSyncRequestHandler - it too `watch`es the peer).

**Comment by AlanVerbner** [comment]
@LukasGasior1 are you planning to include the tests in this PR?

**Comment by LukasGasior1** [comment]
Yes, I'm doing them right now.

**Comment by LukasGasior1** [comment]
@AlanVerbner @adamsmo I've addressed your concerns

**Comment by AlanVerbner** [comment]
ok, will take a second look.

_ps: small thing, please add youtrack id on PR title_

**Comment by adamsmo** [comment]
LGTM

**Comment by adamsmo** [comment]
@LukasGasior1 are we just deleting those tests?
src/test/scala/io/iohk/ethereum/network/p2p/FastSyncActorSpec.scala

File: src/test/scala/io/iohk/ethereum/network/p2p/FastSyncActorSpec.scala [hunk 11]
```diff
@@ -1,278 +0,0 @@
...
-import io.iohk.ethereum.domain.BlockHeader
-import io.iohk.ethereum.network.FastSyncActor.{FastSyncDone, SyncFailure}
-import io.iohk.ethereum.network.PeerActor.MessageReceived
-import io.iohk.ethereum.network.p2p.messages.PV62._
-import io.iohk.ethereum.network.p2p.messages.PV63._
-import io.iohk.ethereum.network.{FastSyncActor, PeerActor}
-import org.scalatest.{FlatSpec, Matchers}
-import org.spongycastle.util.encoders.Hex
-
-class FastSyncActorSpec extends FlatSpec with Matchers {
```
Comment (adamsmo) [inline src/test/scala/io/iohk/ethereum/network/p2p/FastSyncActorSpec.scala]: should we move them sowhere, or implement at least test that checks if fast sync will stop?

Comment (adamsmo) [inline src/test/scala/io/iohk/ethereum/network/p2p/FastSyncActorSpec.scala]: should we move them somewhere, or implement at least test that checks if fast sync will stop?

Comment (LukasGasior1) [inline src/test/scala/io/iohk/ethereum/network/p2p/FastSyncActorSpec.scala]: :+1:

Comment (LukasGasior1) [inline src/test/scala/io/iohk/ethereum/network/p2p/FastSyncActorSpec.scala]: I've added the test for fast sync finish.

File: src/main/scala/io/iohk/ethereum/App.scala (hunk 4 above)
Comment (AlanVerbner) [inline src/main/scala/io/iohk/ethereum/App.scala]: I've been digging around other implementations regarding this and it's definitely for another task

File: src/main/scala/io/iohk/ethereum/blockchain/sync/FastSyncController.scala (hunk 9 above)
Comment (AlanVerbner) [inline src/main/scala/io/iohk/ethereum/blockchain/sync/FastSyncController.scala]: @LukasGasior1 I cannot find where those _block requests_ are returned in `cleanupAndStop` ... i'm expecting something that looks like `fastSyncController ! FastSyncController.EnqueueBlockBodies(requestedHashes)` ... am I missing something?

File: src/main/scala/io/iohk/ethereum/App.scala (hunk 4 above)
Comment (LukasGasior1) [inline src/main/scala/io/iohk/ethereum/App.scala]: @AlanVerbner EC-82 to be precise :)

File: src/main/scala/io/iohk/ethereum/blockchain/sync/FastSyncController.scala (hunk 9 above)
Comment (LukasGasior1) [inline src/main/scala/io/iohk/ethereum/blockchain/sync/FastSyncController.scala]: @AlanVerbner It's in specific request handlers implementations: `FastSyncBlockBodiesRequestHandler`, `FastSyncReceiptsRequestHandler`, etc. See `def handleTimeout` there

Comment (LukasGasior1) [inline src/main/scala/io/iohk/ethereum/blockchain/sync/FastSyncController.scala]: @AlanVerbner well...it will only work for timeout...seems you're right about Terminated

Comment (AlanVerbner) [inline src/main/scala/io/iohk/ethereum/blockchain/sync/FastSyncController.scala]: let me rephrase ... is there any change this actor dies without a timeout?

Comment (LukasGasior1) [inline src/main/scala/io/iohk/ethereum/blockchain/sync/FastSyncController.scala]: yeah, sorry for confusion :) I totally missed that. It needs to be fixed.

Comment (AlanVerbner) [inline src/main/scala/io/iohk/ethereum/blockchain/sync/FastSyncController.scala]: Cool ... sorry for not being explicit on my question

(btw I guess we were both writing at the same time, GH needs live comments updates, lol)

Comment (LukasGasior1) [inline src/main/scala/io/iohk/ethereum/blockchain/sync/FastSyncController.scala]: @AlanVerbner I've just pushed a fix
